### ISSUE TITLE
Styling links as buttons

### ISSUE THREAD

**Issue by jouni** [vaadin/vaadin-lumo-styles#5]
_From @heruan on November 16, 2017 18:7_

I guess this is borderline, but it would be very useful to style links as buttons, e.g. to let the use <kbd>CMD</kbd> + click and open a new tab.

How much the templating system can help on applying the `valo-button` theme module to an anchor element, with all variants, without the need to maintain a duplicate style?

_Copied from original issue: vaadin/vaadin-valo-styles#34_

**Comment by jouni** [comment]
Yeah, would be useful in some cases. I actually wanted to do this in the theme documentation pages, and ended up prototyping this a bit.

It’s possible but requires pretty much 2x selectors in the valo-button style module, so I’m wondering if that’s an acceptable tradeoff?

The way you would use this:

```html
<link rel="import" href="../vaadin-button.html">

<custom-style>
  <style include="valo-button">
  </style>
</custom-style>

<a href="..." theme="button primary large">Button link</a>
```

Examples of the duplication of selectors used in the style module:

```css
:host,
[theme~="button"]


:host(:hover:not([theme*="tertiary"]):not([active])),
[theme~="button"]:hover:not([theme*="tertiary"]):not([active])


:host::before,
:host::after,
[theme~="button"]::before,
[theme~="button"]::after


:host([active]:not([theme~="primary"]):not([theme*="tertiary"])),
[theme~="button"]:active:not([theme~="primary"]):not([theme*="tertiary"])
```

**Comment by jouni** [comment]
_From @heruan on November 24, 2017 17:45_

I didn't understand entirely the duplication of selectors, although it seems acceptable to me. The use case is intuitive and clean! 👍

**Comment by jouni** [comment]
Basically, wherever there’s a `:host` selector, we also need to duplicate it and replace `:host` with `[theme~="button"]`, becuase `:host` won’t match anything outside a custom element.

The `[active]` state needs to be changed to `:active` as well.

**Comment by jouni** [comment]
_From @heruan on January 2, 2018 17:14_

@jouni  Now that the button theme is moved to the button repository, where's the best location to include this (if you're okay to proceed with it, of course!)?

**Comment by jouni** [comment]
This repo would still be the logical place for the styles. `vaadin-button` would then just include the style modules from this package.

I still don’t exactly know how to pull this off nicely. My previous suggestions doesn’t really work, as the `:host` selectors can end up matching unwanted elements (when you import&include the style module in another custom element).

The only other option I now see is to have some build script that generates a `vaadin-button` specific style module from a more generic module. Since there’s one other case that needs a build script (https://github.com/vaadin/vaadin-valo-styles/blob/master/color.html#L190), it might be okay to introduce a build step for this package.

**Comment by heruan** [comment]
Maybe the other way around is easier? See https://github.com/vaadin/vaadin-button/issues/79

**Comment by heruan** [comment]
@jouni Still without any clue on how to do this nicely? I'd really make use of this, for example to trigger a download in Flow you should use an anchor, i.e.

```java
new Anchor("Download me!", streamResource);
```

but placing this along buttons in a toolbar isn't quite pleasant 👊

**Comment by jouni** [comment]
No bright ideas yet.

Have you tried: `<vaadin-button><a>...</a></vaadin-button>`? I bet it has some drawbacks, like having a double tab target (you’d need to set `tabindex=-1` for the link), but it might be workable somehow.

----

I tried one approach in my j-elements research project, how this kind of ”portable CSS” could be used in web components.

Basically, I author the component CSS according to BEM (or close to at least), and the rewrite the selectors when using it inside a shadow root, using the [`bemToShadow` function](https://github.com/jouni/j-elements/blob/master/src/bem-to-shadow.js).

For example, here’s the styles for `<j-avatar>`: https://github.com/jouni/j-elements/blob/master/src/styles/avatar-style.js

And here’s how it’s used in the avatar: [`bemToShadow(style, '.j-avatar')`](https://github.com/jouni/j-elements/blob/master/src/j-avatar.js#L31)
 - `style` is the string from `avatar-style.js`
 - `.j-avatar` is the “block” selector that should be replaced with `:host`

**Comment by AdrienC** [comment]
Hi,
I tried `<a><vaadin-button>...</vaadin-button></a>` to make a RouterLink look like a button, but my whole page was reloaded, so I lost the benefits of my RouterLink.
I tried to style the vaadin-button with 'pointer-events: none' but the button did not have its :hover style and the blur when clicked.
So I disabled pointer-events on the inner `<button>` of the webcomponent, and it works:
```
RouterLink link = new RouterLink("", MainView.class);
Button button = new Button("Home");
button.getElement().setAttribute("tabindex", "-1");
link.add(button);
button.addAttachListener(event -> event.getUI().getPage().executeJavaScript("$0.shadowRoot.lastElementChild.style.pointerEvents = 'none';", button.getElement()));
```

**Comment by Springrbua** [comment]
I just wanted to share my current solution:
I created a custom `Component` like this:

```
public class CustomAnchor extends Composite<Anchor> implements HasTheme, HasStyle {
    /** Custom-Attribut, used to identify the `vaadin-button` as child of `CustomAnchor` */
    private static final String ATTR_ANCHOR        = "anchor";

    private Button btn;

    public CustomAnchor() {
        super();
        btn = new Button();
        btn.setSizeFull();
        btn.getElement().setAttribute(ATTR_ANCHOR, "");
    }
    public CustomAnchor(String href, String text) {
        this();
        setHref(href);
        setText(text);
    }
    // ===============================
    // Other constructors, matching the ones in Anchor
    // ===============================
    protected Anchor initContent() {
        Anchor a = super.initContent();
        a.add(btn);
        return a;
    }
    public void setHref(String href) {
        getContent().setHref(href);
    }
    public void setText(String text) {
        btn.setText(text);
    }

    // Override methods from HasStyle and HasTheme
    // Only methods calling "getElement" have to be overridden

    public void setThemeName(String themeName) {
        btn.setThemeName(themeName);
    }
    // =============================
    // Other methods form HasStyle and HasTheme
    // =============================

    // Support router-link
    public void setRouterLink(boolean value) {
        if (value)
            getElement().setAttribute(ApplicationConstants.ROUTER_LINK_ATTRIBUTE, "");
        else
            getElement().removeAttribute(ApplicationConstants.ROUTER_LINK_ATTRIBUTE);
    }
}
```

Basically I create a `Composite<Anchor>` with a `Button`. In `initContent` this `Button` is added to the `Anchor` as a child.
I implemented `setHref`, which simply calls `setHref` for the `Anchor`.
The method `setText` instead calls `setText` for the `Button`.
To be able to style and theme the `CustomAnchor` the same way you  would style and theme a `Button`, I had to implement `HasStyle` and `HasTheme` and override all of the methods which call `getElement`.
This methods call the respective methods from `Button`.

To support `router-link`, I created a method `setRouterLink(boolean)` which adds / removes the `router-link` attribute.
As mentioned in the comment by @AdrienC the `router-link` is not working by default and the whole page will be reloaded.
Based on his workaround, which disables the `pointer-event` for the inner `<button>`, I created a `dom-module` which does exactly that:

```
<dom-module id="custom-anchor" theme-for="vaadin-button">
    <template>
        <style>
            :host([anchor]) #button {
                pointer-events: none;
            }
        </style>
    </template>
</dom-module>
```

**Comment by Springrbua** [comment]
I just noticed an issue with the solution I posted above:
If I click the label inside the button, the routing works as expected, but if I click outside the label, near the border of the button, the page will be reloaded.
Adding `pointer-events: none` to the `vaadin-button` solves the issue, but then you won't get the hover effects.

**Comment by heruan** [comment]
Thank you @Springrbua for sharing this! I'd like to hear the team's take on it, as it might be worth a PR. Or maybe there's already something on the works with the BEM strategy explained by @jouni?

Sure thing is: styling a link as a button should be easy, as in Bootstrap's `<button class="btn">` and `<a class="btn">`.

**Comment by Springrbua** [comment]
@heruan while my solution works pretty good, I consider it a workaround and I don't think it should become the official solution.
@jouni a possible solution to overcome the duplicate selectors would be to create a custom tag like the vaadin-button does.
The vaadin-button also seems to somehow wrap the native button, the vaadin-anchor could do something similar.
Not sure if thats possible though.

**Comment by jouni** [comment]
One additional point to consider when nesting a button inside an anchor: there should be only one focusable element when navigating with the keyboard. So the nested button should probably have `tabindex=0`. Also, I’m not sure how screen readers interpret this combination.

> @jouni a possible solution to overcome the duplicate selectors would be to create a custom tag like the vaadin-button does.

Yes, that’s one solution, true. Pretty much exactly the same solution as with `<vaadin-button>`, that we want easy-to-style buttons/links. Though, my distaste for this approach has been growing lately. I think I would rather use the native elements directly, and have some additional features on top of them (like shimming the `:focus-visible` pseudo-class) and then some way to include a stylesheet for them in any style scopes that want it.

**Comment by Springrbua** [comment]
> One additional point to consider when nesting a button inside an anchor: there should be only one focusable element when navigating with the keyboard. So the nested button should probably have `tabindex=0`. Also, I’m not sure how screen readers interpret this combination.

Thanks @jouni for the reminder, I did not notice this.
Adding `btn.getElement().setAttribute("tabindex", "-1");` fixed this.

**Comment by kdeda** [comment]
CustomAnchor works perfectly. Thanks.

**Comment by bihu1** [comment]
Based on the previous answer I think the simplest way to solve problem is creating own css class: 
`.reload-disable:active {
	pointer-events: none;
}`

And after that:
`		Button button = new Button();
		button.setClassName("reload-disable");
		RouterLink routerLink = new RouterLink();
		routerLink.add(button);`

In this way hover works perfect.
